nrf24L01_code/bootloader/WriteApp.c

```diff
@@ -0,0 +1,383 @@
+		__nrfRadio_Main();
+		_delay_us(10);
+	}while (!rxData.hasMessage && (timeoutCounter > 0));
+	
+	if (!rxData.hasMessage || (timeoutCounter == 0))
+		rxData.hasMessage = MSG_NO_DATA;		
+}
+
+
+// Bootloader flash
+void boot_program_page (uint32_t page, uint8_t *buf)
+{
+	uint16_t i;
+	uint8_t sreg;
+	// Disable interrupts.
+	sreg = SREG;
+	cli();
+	eeprom_busy_wait ();
+	boot_page_erase (page);
+	boot_spm_busy_wait (); // Wait until the memory is erased.
+	for (i=0; i<SPM_PAGESIZE; i+=2)
+	{
+		// Set up little-endian word.
+		uint16_t w = *buf++;
+		w += (*buf++) << 8;
+		boot_page_fill (page + i, w);
+	}
+	boot_page_write (page); // Store buffer in flash page.
+	boot_spm_busy_wait(); // Wait until the memory is written.
+	// Reenable RWW-section again. We need this if we want to jump back
+	// to the application after bootloading.
+	boot_rww_enable ();
+	// Re-enable interrupts (if they were ever enabled).
+	SREG = sreg;
+}
+
+
+void startFlash(uint8_t * rx_address)
+{
+	uint8_t ackResponse [BUFFER_LENGTH];			
+	uint8_t pageData [PAGE_SIZE];
+	uint8_t bytesReceived = 0;
+	uint16_t currentPage = 0;
+	uint16_t recvCKS = 0;
+
+	do 
+	{
+		switch(bootloader_state)
+		{
+			case BOOTLOADER_INIT_NRF:				
+			{
+				rxData.length = 0x00;
+				rxData.hasMessage = MSG_NO_DATA;
+				initNrf(rx_address);
+				sei();
+				bootloader_state = BOOTLOADER_FLASH_INIT;
+				break;
+			}
+			case BOOTLOADER_FLASH_INIT:
+			{		
+				waitRx();
+				
+				if (rxData.hasMessage)
+				{
+					if (rxData.command == COMM_START_FLASH_PROCEDURE)
+					{
+						bootloader_state = BOOTLOADER_REC_DATA;
+					}
+					else if (rxData.command == COMM_GET_STATE)
+					{
+						memset(ackResponse, 0x00, BUFFER_LENGTH);
+						ackResponse[0] = BOOTLOADER_FLASH_INIT;
+						__nrfRadio_LoadAckPayload(flashPipe, (uint8_t*)ackResponse, BUFFER_LENGTH);
+						
+						waitRx();
+						// We remain in this state
+						bootloader_state = BOOTLOADER_FLASH_INIT;
+					}
+					else
+						bootloader_state = BOOTLOADER_FLASH_ERROR;
+				}
+				else
+				{
+					bootloader_state = BOOTLOADER_FLASH_ERROR;
+				}
+				rxData.hasMessage = MSG_NO_DATA;
+				break;
+			}
+			case BOOTLOADER_REC_DATA:
+			{												
+				do{	
+					waitRx();
+					if (rxData.hasMessage)
+					{
+						if (rxData.command == COMM_SEND_CHUNCK)
+						{							
+							if ((bytesReceived + rxData.length) <= PAGE_SIZE)
+							{	
+								memcpy((uint8_t*) &pageData[bytesReceived],	rxData.data, rxData.length);
+								bytesReceived += rxData.length;
+								rxData.hasMessage = MSG_NO_DATA;
+							}
+							else
+							{								
+								bootloader_state = BOOTLOADER_FLASH_ERROR;
+							}
+						}
+						else if (rxData.command == COMM_GET_STATE)
+						{
+							memset(ackResponse, 0x00, BUFFER_LENGTH);
+							ackResponse[0] = BOOTLOADER_REC_DATA;
+							ackResponse[1] = '0' + bytesReceived / 16;
+							__nrfRadio_LoadAckPayload(flashPipe, (uint8_t*)ackResponse, 2);
+							
+							waitRx();
```

Comment:
this is already called in this loop, makes no sense to call it for a second time.